# Pandas

## Advanced

In [168]:
import numpy as np
import pandas as pd

### Useful functions

#### Operations on dataframe

##### shift

In [169]:
# shift function moves the values up or down by a specified number.
s = pd.Series(np.arange(5)).shift()    # by default down values by 1.
print(s)
print()

s = s.shift(-1)
print(s)

0    NaN
1    0.0
2    1.0
3    2.0
4    3.0
dtype: float64

0    0.0
1    1.0
2    2.0
3    3.0
4    NaN
dtype: float64


##### printing dataframe as string

In [170]:
df = pd.DataFrame(np.array(np.random.randint(1,100,12)).reshape(4,3), columns=list('ABC'), index=list('abcd'))
print(df.to_string())

    A   B   C
a  29  63  31
b  34  56  16
c  60  37   8
d  88  51  74


#### Operations with time

##### Basic creation of time range

In [171]:
# Using date_range
rng = pd.date_range('2025-01-15', periods=5)
print(rng, '\n')

# Using to_datetime
rng = pd.to_datetime([
    '2025-01-15',
    '2025-01-19',
    '2025-03-10',
    '2025-02-16',
    '2025-01-31'
])
print(rng)

DatetimeIndex(['2025-01-15', '2025-01-16', '2025-01-17', '2025-01-18',
               '2025-01-19'],
              dtype='datetime64[us]', freq='D') 

DatetimeIndex(['2025-01-15', '2025-01-19', '2025-03-10', '2025-02-16',
               '2025-01-31'],
              dtype='datetime64[us]', freq=None)


##### freq

In [172]:
# by default = D means 1 day gap
# 2D means 2 days gap
print(pd.date_range('2025-01-15', periods=5, freq='2D'), '\n')

# W for week, ME for month-end, MS for month-start, QE for quarter-end, YE for year-end, h for hour, min for minute, s for second

print(pd.date_range('2025-01-15', periods=5, freq='MS'), '\n')
print(pd.date_range('2025-01-15', periods=5, freq='QE'), '\n')
print(pd.date_range('2025-01-15', periods=5, freq='min'), '\n')

DatetimeIndex(['2025-01-15', '2025-01-17', '2025-01-19', '2025-01-21',
               '2025-01-23'],
              dtype='datetime64[us]', freq='2D') 

DatetimeIndex(['2025-02-01', '2025-03-01', '2025-04-01', '2025-05-01',
               '2025-06-01'],
              dtype='datetime64[us]', freq='MS') 

DatetimeIndex(['2025-03-31', '2025-06-30', '2025-09-30', '2025-12-31',
               '2026-03-31'],
              dtype='datetime64[us]', freq='QE-DEC') 

DatetimeIndex(['2025-01-15 00:00:00', '2025-01-15 00:01:00',
               '2025-01-15 00:02:00', '2025-01-15 00:03:00',
               '2025-01-15 00:04:00'],
              dtype='datetime64[us]', freq='min') 



##### resample

In [173]:
# resample is used to convert timestamped data into a new frequency.
# it works in 2 steps - creates new frequency and then applies aggregation
rng = pd.date_range('2025-03-15', periods=12, freq='5D')
ts = pd.Series(np.random.randint(1,101,12), index=rng)
print(ts[:4])

ts = ts.resample('MS').sum()
ts

2025-03-15    52
2025-03-20    21
2025-03-25    13
2025-03-30    80
Freq: 5D, dtype: int32


2025-03-01    166
2025-04-01    336
2025-05-01     85
Freq: MS, dtype: int32

##### tz_localize and tz_convert

First hh:mm:ss shows the time and second hh:mm shows UTC offset for the timezone.

In [174]:
# tz_localize gives a timezone to a time series.
ts2 = ts.tz_localize('UTC')
ts2

2025-03-01 00:00:00+00:00    166
2025-04-01 00:00:00+00:00    336
2025-05-01 00:00:00+00:00     85
Freq: MS, dtype: int32

In [175]:
# tz_convert converts the timezone to another one and shows information of moments related to older one.
ts3 = ts2.tz_convert('Asia/Kolkata')
ts3

2025-03-01 05:30:00+05:30    166
2025-04-01 05:30:00+05:30    336
2025-05-01 05:30:00+05:30     85
dtype: int32

##### BusinessDay

In [176]:
from pandas.tseries.offsets import BusinessDay
from pandas.tseries.offsets import Day

# BusinessDay skipps the weekends (saturday, sunday) and checks only for working days.
# B in freq of date_range uses frequency for Business Days.
date = pd.Timestamp('2025-01-17')
print(date, date.day_name())

print(date + pd.offsets.Day(1), (date + Day(1)).day_name())
print(date + pd.offsets.BusinessDay(), (date + BusinessDay()).day_name())

2025-01-17 00:00:00 Friday
2025-01-18 00:00:00 Saturday
2025-01-20 00:00:00 Monday


In [177]:
# BusinessDay with series
rng = dates = pd.Series(pd.to_datetime([
    "2025-03-06",
    "2025-03-07",
    "2025-03-08",
    "2025-03-09"
]))
dates + BusinessDay()

0   2025-03-07
1   2025-03-10
2   2025-03-10
3   2025-03-10
dtype: datetime64[us]

#### assign method

##### basic assigning

In [178]:
df = pd.DataFrame({
    "Name": ["A", "B", "C"],
    "Age": [20, 25, 30]
})
df

,Name,Age
0,A,20
1,B,25
2,C,30


In [179]:
df = df.assign(City='Ahmedabad')
df

,Name,Age,City
0,A,20,Ahmedabad
1,B,25,Ahmedabad
2,C,30,Ahmedabad


##### multiple columns

In [180]:
df = df.assign(State='Gujarat', Country='India')
df

,Name,Age,City,State,Country
0,A,20,Ahmedabad,Gujarat,India
1,B,25,Ahmedabad,Gujarat,India
2,C,30,Ahmedabad,Gujarat,India


##### using another column

In [181]:
df = df.assign(copy_age=df['Age'])
print(df)
del df['copy_age']

  Name  Age       City    State Country  copy_age
0    A   20  Ahmedabad  Gujarat   India        20
1    B   25  Ahmedabad  Gujarat   India        25
2    C   30  Ahmedabad  Gujarat   India        30


In [182]:
df = df.assign(next_year_age=df['Age']+1)
df

,Name,Age,City,State,Country,next_year_age
0,A,20,Ahmedabad,Gujarat,India,21
1,B,25,Ahmedabad,Gujarat,India,26
2,C,30,Ahmedabad,Gujarat,India,31


##### boolean assigning

In [183]:
df.assign(Marks=[80,31,92], Result = lambda x: x['Marks'] >= 33)

,Name,Age,City,State,Country,next_year_age,Marks,Result
0,A,20,Ahmedabad,Gujarat,India,21,80,True
1,B,25,Ahmedabad,Gujarat,India,26,31,False
2,C,30,Ahmedabad,Gujarat,India,31,92,True


##### method chaining

In [184]:
df = df.assign(Marks=[80,31,92]).assign(Result = lambda x: x['Marks'] >= 33)
df

,Name,Age,City,State,Country,next_year_age,Marks,Result
0,A,20,Ahmedabad,Gujarat,India,21,80,True
1,B,25,Ahmedabad,Gujarat,India,26,31,False
2,C,30,Ahmedabad,Gujarat,India,31,92,True


### Categoricals

#### Creating a categorical

In [185]:
df = pd.DataFrame({
    'id': np.arange(1,5),
    'grade': ['A', 'B', 'A', 'C']
})
print(df['grade'], '\n')

# Using astype
df['grade'] = df['grade'].astype('category')
print(df['grade'], '\n')

# Using categorical
df['grade'] = pd.Categorical(['A', 'B', 'A', 'C'])
print(df['grade'])

0    A
1    B
2    A
3    C
Name: grade, dtype: str 

0    A
1    B
2    A
3    C
Name: grade, dtype: category
Categories (3, str): ['A', 'B', 'C'] 

0    A
1    B
2    A
3    C
Name: grade, dtype: category
Categories (3, str): ['A', 'B', 'C']


#### Accessing the information of categorical

In [186]:
df['grade'].dtype

CategoricalDtype(categories=['A', 'B', 'C'], ordered=False, categories_dtype=str)

In [187]:
df['grade'].cat.categories

Index(['A', 'B', 'C'], dtype='str')

In [188]:
df['grade'].cat.codes

0    0
1    1
2    0
3    2
dtype: int8

#### Operations on categories

##### Rename the categories

In [189]:
df['grade'] = df['grade'].cat.rename_categories({
    'A': 'Excellent',
    'B': 'Good',
    'C': 'Medium',
})
df['grade']

0    Excellent
1         Good
2    Excellent
3       Medium
Name: grade, dtype: category
Categories (3, str): ['Excellent', 'Good', 'Medium']

##### Set categories

In [190]:
df['grade'] = df['grade'].cat.set_categories(['Excellent', 'Good', 'Medium', "Bad"])
df['grade']

0    Excellent
1         Good
2    Excellent
3       Medium
Name: grade, dtype: category
Categories (4, str): ['Excellent', 'Good', 'Medium', 'Bad']

##### Remove categories

In [191]:
df['grade'] = df['grade'].cat.remove_categories('Bad')
print(df['grade'])

# If you deleted the category which exists in dataframe, NaN will be at location of element having the deleted category.

# Remove unused categories
df['grade'] = df['grade'].cat.remove_unused_categories()

0    Excellent
1         Good
2    Excellent
3       Medium
Name: grade, dtype: category
Categories (3, str): ['Excellent', 'Good', 'Medium']


##### Ordered categories

In [192]:
df['grade'] = df['grade'].cat.set_categories(['Excellent', 'Good', 'Medium', 'Bad'], ordered=True)
print(df['grade'], '\n')

# You can also compare ordered categoricals.
print(df['grade'][0] < df['grade'][1], '\n')
print(df['grade'] > 'Excellent')

0    Excellent
1         Good
2    Excellent
3       Medium
Name: grade, dtype: category
Categories (4, str): ['Excellent' < 'Good' < 'Medium' < 'Bad'] 

True 

0    False
1     True
2    False
3     True
Name: grade, dtype: bool


#### Sorting

In [193]:
df = df.sort_values(by='grade')
df

,id,grade
0,1,Excellent
2,3,Excellent
1,2,Good
3,4,Medium


#### Adding another categories & Reordering categories order

In [194]:
# You can add new categories. They will be stored at last of categories in ordering.
df['grade'] = df['grade'].cat.add_categories('Very Bad')
print(df['grade'], '\n')

# So we have to reorder to store them in order.
df['grade'] = df['grade'].cat.reorder_categories(['Excellent', 'Good', 'Medium', 'Bad', 'Very Bad'])
print(df['grade'])

0    Excellent
2    Excellent
1         Good
3       Medium
Name: grade, dtype: category
Categories (5, str): ['Excellent' < 'Good' < 'Medium' < 'Bad' < 'Very Bad'] 

0    Excellent
2    Excellent
1         Good
3       Medium
Name: grade, dtype: category
Categories (5, str): ['Excellent' < 'Good' < 'Medium' < 'Bad' < 'Very Bad']


#### Grouping

In [195]:
df.groupby('grade', observed=False).size()

grade
Excellent    2
Good         1
Medium       1
Bad          0
Very Bad     0
dtype: int64

### Advanced Concepts

#### Boolean Reduction

##### all

In [196]:
df = pd.DataFrame(np.arange(12).reshape(4,3), columns=list('ABC'), index=list('abcd'))

# all() checks whether condition is true for all the elements or not of a row or a column.
print((df > 1).all(), '\n')         # by default, axis=0
print((df > 3).all(axis=1))

A    False
B    False
C     True
dtype: bool 

a    False
b    False
c     True
d     True
dtype: bool


##### any

In [197]:
# any() checks whether condition is true for any of the elements or not of a row or a column.
print((df > 1).any(), '\n')
print((df < 1).any())

A    True
B    True
C    True
dtype: bool 

A     True
B    False
C    False
dtype: bool


##### empty

In [198]:
df1 = pd.DataFrame()
df2 = pd.DataFrame(columns=list('ABC'), index=list('abcd'))

# empty checks is dataframe empty or not.
print(df1.empty)
print(df2.empty)

True
False


#### Binary comparison methods

In [199]:
# Binary comparison methods compares two operands and returns True or False.
df = pd.DataFrame(np.arange(12).reshape(4,3), columns=list('ABC'), index=list('abcd'))
print(df, '\n')

print(df.eq(5), '\n')
print(df.le(5))

# eq : equals to : ==
# nq : not equals to : !=
# lt : lesser than : <
# gt : greater than : >
# le : lesser than or equals to : <=
# ge : greater than or equals to : >=

   A   B   C
a  0   1   2
b  3   4   5
c  6   7   8
d  9  10  11 

       A      B      C
a  False  False  False
b  False  False   True
c  False  False  False
d  False  False  False 

       A      B      C
a   True   True   True
b   True   True   True
c  False  False  False
d  False  False  False


### Console display

#### set_option

##### width

In [200]:
df1 = pd.DataFrame(np.array(np.random.randint(10000000,10000000000, 50, dtype='i8')).reshape(2,25))

# pd.set_option('display.width', 500) This statement will set changes permanently for all outputs.
# To change options temperorily, do as below.

with pd.option_context('display.width', 150):
    print(df1)

           0           1           2           3           4           5           6           7           8           9   ...          15  \
0  6058885216  7071695780  8681468223  2151310495  9422106547  6282618976  7391774458  1017118089  1992253178  7383442272  ...  2470839112   
1   782768041  5849918053  7179170566  2797894356  7569088825  4461356321  4845248095  1094866422  7588187488  8205830082  ...  7302201831   

           16          17          18          19          20          21          22          23          24  
0  2199016088  5955472967   708791856  2958265822  1714251704   526029688  6208728177  9922518990   876802067  
1  3115426416  9173736953  2652611088  7050889408  9235611748  9535238915  5028134466  5201575563  4182946271  

[2 rows x 25 columns]


##### max_colwidth

In [201]:
df2 = pd.DataFrame({
    "Name": ["Dhvanit", "Rahul"],
    "Description": [
        "This is a very long description about the first person",
        "This is another very long description about the second person"
    ]
})

with pd.option_context("display.max_colwidth", 20):
    print(df2)

      Name          Description
0  Dhvanit  This is a very l...
1    Rahul  This is another ...


##### max_columns

In [202]:
with pd.option_context('display.max_columns', 15):
    print(df1)

           0           1           2           3           4           5   \
0  6058885216  7071695780  8681468223  2151310495  9422106547  6282618976   
1   782768041  5849918053  7179170566  2797894356  7569088825  4461356321   

           6   ...          18          19          20          21  \
0  7391774458  ...   708791856  2958265822  1714251704   526029688   
1  4845248095  ...  2652611088  7050889408  9235611748  9535238915   

           22          23          24  
0  6208728177  9922518990   876802067  
1  5028134466  5201575563  4182946271  

[2 rows x 25 columns]


##### max_rows

In [203]:
df3 = pd.DataFrame(np.arange(30).reshape(15,2))

with pd.option_context('display.max_rows', 10):
    print(df3)

     0   1
0    0   1
1    2   3
2    4   5
3    6   7
4    8   9
..  ..  ..
10  20  21
11  22  23
12  24  25
13  26  27
14  28  29

[15 rows x 2 columns]


##### expand_frame_repr

In [204]:
# If it's True, output of dataframe can be wrapped if it's not fit in allowed width.
# If it's false, output will be strictly printed in a single horizontal line.

with pd.option_context(
    'display.width', 150,
    'display.expand_frame_repr', True
):
    print(df1)

print('\n----------------------------------------------------------------------------------------------------------------------------------------\n')

with pd.option_context(
    'display.width', 150,
    'display.expand_frame_repr', False
):
    print(df1)

           0           1           2           3           4           5           6           7           8           9   ...          15  \
0  6058885216  7071695780  8681468223  2151310495  9422106547  6282618976  7391774458  1017118089  1992253178  7383442272  ...  2470839112   
1   782768041  5849918053  7179170566  2797894356  7569088825  4461356321  4845248095  1094866422  7588187488  8205830082  ...  7302201831   

           16          17          18          19          20          21          22          23          24  
0  2199016088  5955472967   708791856  2958265822  1714251704   526029688  6208728177  9922518990   876802067  
1  3115426416  9173736953  2652611088  7050889408  9235611748  9535238915  5028134466  5201575563  4182946271  

[2 rows x 25 columns]

----------------------------------------------------------------------------------------------------------------------------------------

           0           1           2           3           4           5   

#### get_option

In [205]:
# We can print values of any display option using get_option.

print(pd.get_option("display.width"))
print(pd.get_option("display.max_columns"))
print(pd.get_option("display.max_rows"))
print(pd.get_option("display.max_colwidth"))
print(pd.get_option("display.expand_frame_repr"))

80
20
60
50
True


### Discretization and quantiling

#### cut()

##### Basic cut() operation

cut() divides the range from minimum to maximum into given number of intervals.

In [206]:
s = pd.Series(np.array([1,4,5,11,7]))
pd.cut(s, bins=3)

0     (0.99, 4.333]
1     (0.99, 4.333]
2    (4.333, 7.667]
3     (7.667, 11.0]
4    (4.333, 7.667]
dtype: category
Categories (3, interval[float64, right]): [(0.99, 4.333] < (4.333, 7.667] < (7.667, 11.0]]

In output, cut() will assign values depending on comparison between intervals value and data's value.
<br>
Above example :
<br>
3 intervals will be created between 1(minimum) and 11(maximum), it's nearby 3.33. So intervals are among 1, 4.33, 7.67, 11(First value excluded).
<br>
1  -> 1 is lesser than 4.33 so assigned by (1, 4.33).
<br>
4  -> 4 is lesser than 4.33 so assigned by (1, 4.33).
<br>
5  -> 5 is between 4.33 and 7.67 so assigned by (4.33, 7.67).
<br>
11 -> 11 is greater than 7.67 so assigned by (7.67, 11).
<br>
7  -> 7 is between 4.33 and 7.67 so assigned by (4.33, 7.67).
<br>
If a data's value is exactly matched with interval's value(i.e : np.array([1,4.333,5,11,7])), the previous interval value will be assigned to it.

0.99 is a small padding applied by pandas to the min value of data so min value can be included in first interval.
<br>
If pandas doesn't apply this padding and keeps min value in first interval, min value will be excluded from the first interval.

##### specifying own bins

In [207]:
# We can define the boundries.
pd.cut(s, bins=[0, 2.5, 5, 7.5, 10, 12.5])

0      (0.0, 2.5]
1      (2.5, 5.0]
2      (2.5, 5.0]
3    (10.0, 12.5]
4      (5.0, 7.5]
dtype: category
Categories (5, interval[float64, right]): [(0.0, 2.5] < (2.5, 5.0] < (5.0, 7.5] < (7.5, 10.0] < (10.0, 12.5]]

##### giving labels

In [208]:
pd.cut(s, bins=3, labels=['low', 'medium', 'high'])

0       low
1       low
2    medium
3      high
4    medium
dtype: category
Categories (3, str): ['low' < 'medium' < 'high']

##### right parameter

In [209]:
# By default left value of interval is excluded and right value of interval is included.
# We can change this setting to opposite using right parameter.

pd.cut(s, bins=3, right=False)

0      [1.0, 4.333)
1      [1.0, 4.333)
2    [4.333, 7.667)
3    [7.667, 11.01)
4    [4.333, 7.667)
dtype: category
Categories (3, interval[float64, left]): [[1.0, 4.333) < [4.333, 7.667) < [7.667, 11.01)]

##### include_lowest

In [210]:
# We can define that the minimum value should be included in first interval even if it's exactly same with first interval's first value and settings are set to exclude let value.
print(pd.cut(s, bins=[1,4,7,11]), '\n')
print(pd.cut(s, bins=[1,4,7,11], include_lowest=True))

0            NaN
1     (1.0, 4.0]
2     (4.0, 7.0]
3    (7.0, 11.0]
4     (4.0, 7.0]
dtype: category
Categories (3, interval[int64, right]): [(1, 4] < (4, 7] < (7, 11]] 

0    (0.999, 4.0]
1    (0.999, 4.0]
2      (4.0, 7.0]
3     (7.0, 11.0]
4      (4.0, 7.0]
dtype: category
Categories (3, interval[float64, right]): [(0.999, 4.0] < (4.0, 7.0] < (7.0, 11.0]]


##### getting the bin boundaries

In [211]:
pd.cut(s, bins=[0,4,7,11], retbins=True)

(0     (0, 4]
 1     (0, 4]
 2     (4, 7]
 3    (7, 11]
 4     (4, 7]
 dtype: category
 Categories (3, interval[int64, right]): [(0, 4] < (4, 7] < (7, 11]],
 array([ 0,  4,  7, 11]))

#### qcut()

##### Basic qcut() operation

In [212]:
# qcut divides the series to parts(quantiles) having equal number of values.
# qcut defines boundaries correspoding to values of series instead of range.
s = pd.Series(np.arange(1,10))
pd.qcut(s, q=3)

0    (0.999, 3.667]
1    (0.999, 3.667]
2    (0.999, 3.667]
3    (3.667, 6.333]
4    (3.667, 6.333]
5    (3.667, 6.333]
6      (6.333, 9.0]
7      (6.333, 9.0]
8      (6.333, 9.0]
dtype: category
Categories (3, interval[float64, right]): [(0.999, 3.667] < (3.667, 6.333] < (6.333, 9.0]]

##### specifying own quantiles

In [213]:
# We can define the quantity of each quantiles.
pd.qcut(s, q=[0, 0.2, 0.8, 1])

0    (0.999, 2.6]
1    (0.999, 2.6]
2      (2.6, 7.4]
3      (2.6, 7.4]
4      (2.6, 7.4]
5      (2.6, 7.4]
6      (2.6, 7.4]
7      (7.4, 9.0]
8      (7.4, 9.0]
dtype: category
Categories (3, interval[float64, right]): [(0.999, 2.6] < (2.6, 7.4] < (7.4, 9.0]]

##### labels

In [214]:
pd.qcut(s, q=3, labels=['Q1','Q2','Q3'])

0    Q1
1    Q1
2    Q1
3    Q2
4    Q2
5    Q2
6    Q3
7    Q3
8    Q3
dtype: category
Categories (3, str): ['Q1' < 'Q2' < 'Q3']

##### getting the bin boundaries

In [215]:
pd.qcut(s, q=2, retbins=True)

(0    (0.999, 5.0]
 1    (0.999, 5.0]
 2    (0.999, 5.0]
 3    (0.999, 5.0]
 4    (0.999, 5.0]
 5      (5.0, 9.0]
 6      (5.0, 9.0]
 7      (5.0, 9.0]
 8      (5.0, 9.0]
 dtype: category
 Categories (2, interval[float64, right]): [(0.999, 5.0] < (5.0, 9.0]],
 array([1., 5., 9.]))

##### duplicate values

In [216]:
s2 = pd.Series(np.array([10,20,20,10,20,10]))
# pd.qcut(s2, q=4) This statement will produce an error because the data cannot be divided into 4 parts having quplicate values.
# So below statement will remove the duplicate bin edges instead of raising an error.
pd.qcut(s2, q=4, duplicates='drop')

0    (9.999, 15.0]
1     (15.0, 20.0]
2     (15.0, 20.0]
3    (9.999, 15.0]
4     (15.0, 20.0]
5    (9.999, 15.0]
dtype: category
Categories (2, interval[float64, right]): [(9.999, 15.0] < (15.0, 20.0]]

#### Practical Usage

In [217]:
df = pd.DataFrame({
    "customer": ["A", "B", "C", "D", "E", "F"],
    "spending": [100, 250, 300, 450, 700, 1000]
})

df["category"] = pd.cut(df["spending"], bins=[0, 300, 700, 1000], labels=["Low", "Medium", "High"])
df['category']

# We can analyze behavior of values by seeing their existance in part of range of data.

0       Low
1       Low
2       Low
3    Medium
4    Medium
5      High
Name: category, dtype: category
Categories (3, str): ['Low' < 'Medium' < 'High']

In [218]:
df["category"] = pd.qcut(
    df["spending"],
    q=3,
)
df['category']

# We can analyze behavior of rate of changing values by seeing range of quantiles of data.

0     (99.999, 283.333]
1     (99.999, 283.333]
2    (283.333, 533.333]
3    (283.333, 533.333]
4     (533.333, 1000.0]
5     (533.333, 1000.0]
Name: category, dtype: category
Categories (3, interval[float64, right]): [(99.999, 283.333] < (283.333, 533.333] < (533.333, 1000.0]]